### PR TITLE
WIP: Testing Assert improve

### PR DESCRIPTION
**Pull request by zekth**
At the moment i have added 2 assert functions
- stringContains // strictly using string.includes()
- stringMatch // using regexp

I have some questions:
- For stringContains, do we add a parameter for `case-insensitive` ?
- Wanted to add Object equals (properties not reference) and Matches. Do you want me to add it?
- Wanted to add arrayContains (which requires also Object equals). Do you want me to add it?

@hayd your thoughts?

### CODE REVIEW

**Comment by ry** [comment]
Cool - looks good. 
Can you move the asserts out of testing/mod.ts and into testing/asserts.ts ?

I don't feel so happy about having this node-style assert library.. I feel like it's not in the same style as the rest of deno_std... Would prefer to have something like

```ts
import { assertEqual } from "https://deno.land/x/testing/assert.ts";
```

**Comment by zekth** [comment]
Sure seems totally legit. Will refactor and add features so.

**Comment by ry** [comment]
Please rebase

**Comment by zekth** [comment]
aybabtu

**Comment by zekth** [comment]
@ry i have moved all the assertions into the assert.ts. I also have refactored diffType in an enum instead of type. Seems more intellegible.

But i have a question about one exported function of the module:
https://github.com/denoland/deno_std/blob/caf6876c59214635c1c216bf4ec662cb42fe63dc/testing/mod.ts#L16

This one is exposed and used but is it deprecated as now we have to only use the pretty.ts ? Do i deprecate it and migrate all the UTs ?

File: testing/assert.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  if (!expr) {
+    throw new Error(msg);
+  }
+}
+
+/**
+ * Make an assertion that `actual` and `expected` are equal, deeply. If not
+ * deeply equal, then throw.
+ */
+export function equal(actual: unknown, expected: unknown, msg?: string): void {
```
Comment (ry) [inline testing/assert.ts]: Rename to `assertEq`

File: testing/assert.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  prettyAssertEqual(actual, expected, msg);
+}
+
+/**
+ * Make an assertion that `actual` and `expected` are strictly equal.  If
+ * not then throw.
+ */
+export function strictEqual(
```
Comment (ry) [inline testing/assert.ts]: Rename to `assertStrictEq`

File: testing/assert.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    throw new Error(msg);
+  }
+}
+
+/**
+ * Make an assertion that actual contains expected. If not
+ * then thrown.
+ */
+export function stringContains(
```
Comment (ry) [inline testing/assert.ts]: Rename to `assertStringContains`

File: testing/assert.ts (hunk 1 above)
Comment (zekth) [reply testing/assert.ts]: I only rename in this module not in the mod.ts `const assertions = { .... ` ? right?

File: testing/assert.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    throw new Error(msg);
+  }
+}
+
+/**
+ * Make an assertion that `actual` match RegExp `expected`. If not
+ * then thrown
+ */
+export function stringMatch(
```
Comment (ry) [inline testing/assert.ts]: rename to `assertStringMatch`

File: testing/assert.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+export function fail(msg?: string): void {
+  // eslint-disable-next-line @typescript-eslint/no-use-before-define
+  assert(false, `Failed assertion${msg ? `: ${msg}` : "."}`);
+}
+
+/** Executes a function, expecting it to throw.  If it does not, then it
+ * throws.  An error class and a string that should be included in the
+ * error message can also be asserted.
+ */
+export function throws(
```
Comment (ry) [inline testing/assert.ts]: Rename to `assertThrows`

File: testing/assert.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    doesThrow = true;
+  }
+  if (!doesThrow) {
+    msg = `Expected function to throw${msg ? `: ${msg}` : "."}`;
+    throw new Error(msg);
+  }
+}
+
+export async function throwsAsync(
```
Comment (ry) [inline testing/assert.ts]: rename to `assertThrowsAsync`

File: testing/assert.ts (hunk 1 above)
Comment (ry) [reply testing/assert.ts]: @zekth Sure keep the existing API in testing/mod.ts for now... but I'm going to remove that too and require people to import from here. But maybe that's better done in a separate commit.

File: testing/assert.ts (hunk 3 above)
Comment (ry) [reply testing/assert.ts]: Maybe `assertHasSubStr`

File: testing/assert.ts (hunk 1 above)
Comment (zekth) [reply testing/assert.ts]: agreed.

File: testing/assert.ts (hunk 3 above)
Comment (ry) [reply testing/assert.ts]: Maybe `assertContains`

Comment (zekth) [reply testing/assert.ts]: strContains seems more legitimate because i wanted to add arrContains assert after.

File: testing/assert.ts [hunk 7]
```diff
@@ -25,7 +25,7 @@ export function equal(actual: unknown, expected: unknown, msg?: string): void {
  * Make an assertion that `actual` and `expected` are strictly equal.  If
```
Comment (ry) [inline testing/assert.ts]: Rename this file testing/asserts.ts and the other testing/asserts_test.ts

File: testing/bench.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { bench, runBenchmarks } from "./../benching/mod.ts";
 import { runTests } from "./mod.ts";
 
-import "./test.ts";
+import "./asserts_test.ts";
```
Comment (ry) [inline testing/bench.ts]: This line can be removed completely... (I think)

File: testing/mod.ts [hunk 9]
```diff
@@ -36,125 +41,14 @@ export function equal(c: unknown, d: unknown): boolean {
...
-    }
-  }
+  assert: assertImport,
+  equal: AssertEqual,
+  strictEqual: assertStrictEq,
+  assertStrContains: assertStrContains,
+  assertMatch: assertMatch,
+  fail: fail,
+  throws: assertThrows,
+  throwsAsync: assertThrowsAsync
```
Comment (ry) [inline testing/mod.ts]: 👍

**Pull request merged**